ARROW-15972: [Java][Doc] Add Getting Started section (#12739)
apache/arrow: docs/source/java/quickstartguide.rst

Example: Read the dataset from the previous example to an Arrow random-access file.

.. code-block:: Java

    import org.apache.arrow.memory.RootAllocator;
    import org.apache.arrow.vector.ipc.ArrowFileReader;
    import org.apache.arrow.vector.ipc.message.ArrowBlock;
    import org.apache.arrow.vector.VectorSchemaRoot;
    import java.io.File;
    import java.io.FileInputStream;
    import java.io.FileOutputStream;
    import java.io.IOException;

    try(
        BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
        FileInputStream fileInputStream = new FileInputStream(new File("random_access_file.arrow"));
        ArrowFileReader reader = new ArrowFileReader(fileInputStream.getChannel(), allocator);
    ){
        System.out.println("Record batches in file: " + reader.getRecordBlocks().size());
        for (ArrowBlock arrowBlock : reader.getRecordBlocks()) {
            reader.loadRecordBatch(arrowBlock);
            VectorSchemaRoot root = reader.getVectorSchemaRoot();
            System.out.println("VectorSchemaRoot read: \n" + root.contentToTSVString());
        }
    } catch (IOException e) {
        e.printStackTrace();
    }

.. code-block:: shell

>     VectorSchemaRoot read:

Reviewer: This output is missing the "Record batches in file" line

Author: Added

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. default-domain:: java
.. highlight:: java

=================
Quick Start Guide
=================

.. contents::

Arrow Java provides several building blocks. Data types describe the types of values;

> ``ValueVectors`` are sequences of typed values; ``fields`` describe the types of columns in

Reviewer: We don't need to put any of these terms in a code font

Author: Changed

tabular data; ``schemas`` describe a sequence of columns in tabular data, and
``VectorSchemaRoot`` represents tabular data. Arrow also provides ``readers`` and
``writers`` for loading data from and persisting data to storage.

> Create a ValueVector
> *********************

Reviewer: ```suggestion
Create a ValueVector
********************
```

    }

.. code-block:: shell

    Vector created in memory: [one, two, three]

Create a Field
**************

**Fields** are used to denote the particular columns of tabular data.
They consist of a name, a data type, a flag indicating whether the column can have null values,
and optional key-value metadata.

Example: create a field named "document" of string type:

.. code-block:: Java

    import org.apache.arrow.vector.types.pojo.ArrowType;
    import org.apache.arrow.vector.types.pojo.Field;
    import org.apache.arrow.vector.types.pojo.FieldType;
    import java.util.HashMap;
    import java.util.Map;

    Map<String, String> metadata = new HashMap<>();
    metadata.put("A", "Id card");
    metadata.put("B", "Passport");
    metadata.put("C", "Visa");
    Field document = new Field("document",
            new FieldType(true, new ArrowType.Utf8(), /*dictionary*/ null, metadata),
            /*children*/ null);
    System.out.println("Field created: " + document + ", Metadata: " + document.getMetadata());

.. code-block:: shell

    Field created: document: Utf8, Metadata: {A=Id card, B=Passport, C=Visa}

Create a Schema
***************

> **Schema** holds a sequence of fields together with some optional metadata.

Reviewer: ```suggestion
**Schemas** hold a sequence of fields together with some optional metadata.
```

Example: Create a schema describing datasets with two columns:

> a int32 column "A" and a utf8-encoded string column "B"

Reviewer: ```suggestion
an int32 column "A" and a UTF8-encoded string column "B"
```

.. code-block:: Java

    import org.apache.arrow.vector.types.pojo.ArrowType;
    import org.apache.arrow.vector.types.pojo.Field;
    import org.apache.arrow.vector.types.pojo.FieldType;
    import org.apache.arrow.vector.types.pojo.Schema;
    import java.util.HashMap;
    import java.util.Map;
    import static java.util.Arrays.asList;

    Map<String, String> metadata = new HashMap<>();
    metadata.put("K1", "V1");
    metadata.put("K2", "V2");
    Field a = new Field("A", FieldType.nullable(new ArrowType.Int(32, true)), /*children*/ null);
    Field b = new Field("B", FieldType.nullable(new ArrowType.Utf8()), /*children*/ null);
    Schema schema = new Schema(asList(a, b), metadata);
    System.out.println("Schema created: " + schema);

.. code-block:: shell

    Schema created: Schema<A: Int(32, true), B: Utf8>(metadata: {K1=V1, K2=V2})

Create a VectorSchemaRoot
*************************

> **VectorSchemaRoot** combines ValueVectors with a Schema to represent tabular data.

Reviewer: ```suggestion
A **VectorSchemaRoot** combines ValueVectors with a Schema to represent tabular data.
```

> Example: Create a dataset with metadata that contains integer age and
> string names of data.

Reviewer: ```suggestion
Example: Create a dataset of names (strings) and ages (32-bit signed integers).
```

.. code-block:: Java

    import org.apache.arrow.memory.BufferAllocator;
    import org.apache.arrow.memory.RootAllocator;
    import org.apache.arrow.vector.IntVector;
    import org.apache.arrow.vector.VarCharVector;
    import org.apache.arrow.vector.VectorSchemaRoot;
    import org.apache.arrow.vector.types.pojo.ArrowType;
    import org.apache.arrow.vector.types.pojo.Field;
    import org.apache.arrow.vector.types.pojo.FieldType;
    import org.apache.arrow.vector.types.pojo.Schema;
    import java.nio.charset.StandardCharsets;
    import java.util.HashMap;
    import java.util.Map;
    import static java.util.Arrays.asList;

>     Field a = new Field("age",

Reviewer: ```suggestion
    Field age = new Field("age",
```

            FieldType.nullable(new ArrowType.Int(32, true)),

>             /*children*/null

Reviewer: Does Field require you to pass children?

Author: Yes, it is required

    );

>     Field b = new Field("name",

Reviewer: ```suggestion
    Field name = new Field("name",
```

>             new FieldType(true, new ArrowType.Utf8(), /*dictionary*/ null, /*metadata*/ null),

Reviewer: Can we use `FieldType.nullable` here?

Author: Changed

            /*children*/null
    );
    Schema schema = new Schema(asList(a, b), /*metadata*/ null);
    try(
        BufferAllocator allocator = new RootAllocator();
        VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator);

>         IntVector intVectorA = (IntVector) root.getVector("age");
>         VarCharVector varCharVectorB = (VarCharVector) root.getVector("name");

Reviewer: ```suggestion
        IntVector ageVector = (IntVector) root.getVector("age");
        VarCharVector nameVector = (VarCharVector) root.getVector("name");
```

Arrow data can be written to and read from disk, and both of these can be done in
a streaming and/or random-access fashion depending on application requirements.

**Write data to an arrow file**

Example: Write the dataset from the previous example to an Arrow random-access file.

.. code-block:: Java

    import org.apache.arrow.memory.BufferAllocator;
    import org.apache.arrow.memory.RootAllocator;
    import org.apache.arrow.vector.IntVector;
    import org.apache.arrow.vector.VarCharVector;
    import org.apache.arrow.vector.VectorSchemaRoot;
    import org.apache.arrow.vector.ipc.ArrowFileWriter;
    import org.apache.arrow.vector.types.pojo.ArrowType;
    import org.apache.arrow.vector.types.pojo.Field;
    import org.apache.arrow.vector.types.pojo.FieldType;
    import org.apache.arrow.vector.types.pojo.Schema;
    import java.io.File;
    import java.io.FileOutputStream;
    import java.io.IOException;
    import java.nio.charset.StandardCharsets;
    import java.util.HashMap;
    import java.util.Map;
    import static java.util.Arrays.asList;

    Map<String, String> metadataField = new HashMap<>();
    metadataField.put("K1-Field", "K1F1");

>     metadataField.put("K2-Field", "K2F2");

Reviewer: IMO, we can remove the metadata here to keep this example more manageable.

Author: Deleted

>     Field a = new Field("Column-A-Age",

Reviewer: ```suggestion
    Field age = new Field("age",
```

Reviewer: and similar below

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

.. default-domain:: java
.. highlight:: java

> .. _java_quickstartguide:

Reviewer: note it's not too useful to have this anchor, 1) we don't use it and 2) for linking to a file, instead of a subheading, just use `:doc:`

Author: Deleted

=================

> Quick Start Guide

Author: Changed

=================

.. contents::

Arrow java manage data in Vector Schema Root (somewhat analogous to tables and record

> batches in the other Arrow implementations). Before to create a Vector Schema Root let's

Reviewer: We should probably just consistently use "VectorSchemaRoot"

Reviewer: And we can introduce things more naturally, something like: Arrow provides several building blocks; ValueVectors are sequences of typed values; data types, fields, and schemas describe the types of values, and the types of columns in tabular data; VectorSchemaRoot represents tabular data. Arrow also provides readers and writers for loading data from and persisting data to storage.

define another topics neededs for that purpose.

Create A Value Vector
*********************

> It's called `array` in the columnar format specification. Represent a one-dimensional

Reviewer: reST uses double backticks. But usually for this we should use quotes. `Also known as "arrays" in the columnar format.`

Author: Added

    metadataSchema.put("K2-Schema", "K2S2");
    Schema schema = new Schema(asList(a, b), metadataSchema);
    System.out.println("Field A: " + a);
    System.out.println("Field B: " + b + ", Metadata: " + b.getMetadata());
    System.out.println("Schema: " + schema);
    try(BufferAllocator rootAllocator = new RootAllocator();
        VectorSchemaRoot vectorSchemaRoot = VectorSchemaRoot.create(schema, rootAllocator)){
        vectorSchemaRoot.setRowCount(3);
        try(IntVector intVectorA = (IntVector) vectorSchemaRoot.getVector("Column-A-Age");
            VarCharVector varCharVectorB = (VarCharVector) vectorSchemaRoot.getVector("Column-B-Name")) {
            intVectorA.allocateNew(3);
            intVectorA.set(0, 10);
            intVectorA.set(1, 20);
            intVectorA.set(2, 30);

            varCharVectorB.allocateNew(3);
            varCharVectorB.set(0, "Dave".getBytes(StandardCharsets.UTF_8));
            varCharVectorB.set(1, "Peter".getBytes(StandardCharsets.UTF_8));
            varCharVectorB.set(2, "Mary".getBytes(StandardCharsets.UTF_8));

            System.out.println("Vector Schema Root: \n" + vectorSchemaRoot.contentToTSVString());
        }
    }

> Create a IPC File or Random Access Format

Author: Added

*****************************************

The Arrow Interprocess Communication (IPC) format defines two types of binary
formats for serializing Arrow data: the streaming format and the file format
(or random access format). Such files can be directly memory-mapped when read.

**Write File or Random Access Format**: Write to a file a dataset with metadata

> that contains integer age and string names of data.

Reviewer: I think we can be briefer here and just say that Arrow data can be written to and read from disk, and both of these can be done in a streaming and/or random-access fashion depending on application requirements.

Author: Added

            intVectorA.allocateNew(3);
            intVectorA.set(0, 10);
            intVectorA.set(1, 20);
            intVectorA.set(2, 30);
            varCharVectorB.allocateNew(3);
            varCharVectorB.set(0, "Dave".getBytes(StandardCharsets.UTF_8));
            varCharVectorB.set(1, "Peter".getBytes(StandardCharsets.UTF_8));
            varCharVectorB.set(2, "Mary".getBytes(StandardCharsets.UTF_8));
            // Arrow Java At Rest
            File file = new File("randon_access_to_file.arrow");
            try (FileOutputStream fileOutputStream = new FileOutputStream(file);
                 ArrowFileWriter writer = new ArrowFileWriter(vectorSchemaRoot, null, fileOutputStream.getChannel())
            ) {
                writer.start();
                writer.writeBatch();
                writer.end();
                System.out.println("Record batches written: " + writer.getRecordBlocks().size() + ". Number of rows written: " + vectorSchemaRoot.getRowCount());
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

> Create A Flight Server And Flight Client

Reviewer: IMO this section can be omitted, we can defer this for a dedicated page about Flight

Author: Deleted
